### PR TITLE
Implement shred

### PR DESCRIPTION
**Pull request by forticulous**
This is my attempt at implementing shred.  It's partially based on some earlier implementation I saw in the issues section that someone else had worked on.  Feedback appreciated

### CODE REVIEW

**Comment by ebfe** [comment]
This should have a warning that it is pretty much uselss on every modern filesystem and likely on SSDs too regardless of the filesystem in use.

Instead of File::create repeatedly it probably should File::open it only once (ideally with O_DIRECT|O_SYNC).

File: src/shred/shred.rs [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+        Err(_) => {
+            println!("{}: {}: Invalid file size", NAME, size_str_opt.unwrap());
+            exit!(1);
+        }
+    };
+    
+    Some(coeff * unit)
+}
+
+fn bytes_to_string(bytes: &[u8]) -> String {
```
Comment (ebfe) [inline src/shred/shred.rs]: This looks strange. What are you trying to achive here?

File: src/shred/shred.rs [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+                }
+                Err(e) => {
+                    println!("{}: {}: Couldn't rename to {}: {}", NAME,
+                                                                  last_path.display(),
+                                                                  new_path.display(),
+                                                                  e);
+                    return None;
+                }
+            }
+        } // If every possible filename already exists, just reduce the length and try again
```
Comment (ebfe) [inline src/shred/shred.rs]: There is a race condition between checking new_path.exists() and fs::rename(). new_path might have been created between the exists() call and the rename() call. Just check the error returned by rename() if it is necessary to distinguish errors.

File: src/shred/shred.rs [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+use std::io;
+use std::io::prelude::*;
+use std::path::{Path, PathBuf};
+
+#[macro_use]
+extern crate uucore;
+
+static NAME: &'static str = "shred";
+static VERSION_STR: &'static str = "1.0.0";
+const BLOCK_SIZE: usize = 512;
```
Comment (ebfe) [inline src/shred/shred.rs]: Probably a good default, but why not fs::Metadata::blksize()? Not sure it matters.

File: src/shred/shred.rs (hunk 1 above)
Comment (forticulous) [inline src/shred/shred.rs]: Looks like it's just parsing the size argument into a u64.  What part looks strange?

File: src/shred/shred.rs (hunk 3 above)
Comment (forticulous) [inline src/shred/shred.rs]: I looked into MetadataExt::blksize(), it looks like it's only for unix file systems

File: src/shred/shred.rs (hunk 1 above)
Comment (ebfe) [inline src/shred/shred.rs]: I mean the bytes_to_string function

File: src/shred/shred.rs (hunk 2 above)
Comment (forticulous) [inline src/shred/shred.rs]: You're right that the file could have been created between the exists() and rename() calls, but I'm not sure how to handle the scenario.  I made a little test program on my machine that renames a file to an existing path and it just silently overwrote the file.  I wouldn't be able to determine that it happened by looking for errors

File: src/shred/shred.rs (hunk 1 above)
Comment (forticulous) [inline src/shred/shred.rs]: Oh that is a kind of weird function.  It just wants to create a string to print out for what type of pass just occurred like when a zero pass happens it should show "pass 1/2 (000000)...".  Although the byte pattern can be something other than the zero byte I think in practice it never will be

Comment (forticulous) [inline src/shred/shred.rs]: In the latest commit I removed the File::create calls because I realized it truncates existing files by default

Comment (ebfe) [inline src/shred/shred.rs]: > > +fn bytes_to_string(bytes: &[u8]) -> String {
> 
> Oh that is a kind of weird function.  It just wants to create a string to print out for what type of pass just occurred like when a zero pass happens it should show "pass 1/2 (000000)...".  Although the byte pattern can be something other than the zero byte I think in practice it never will be

Well it is. Try `shred -n 10 x`.

**Comment by ebfe** [comment]
Also since this seems to be largely based on #527, you shouldn't remove
the original authors copyright line.

File: src/shred/shred.rs (hunk 3 above)
Comment (ebfe) [inline src/shred/shred.rs]: > > +const BLOCK_SIZE: usize = 512;
> 
> I looked into MetadataExt::blksize(), it looks like it's only for unix file systems

Does shred do anything useful on NTFS? If not we can just make it a unix
only util.

**Comment by forticulous** [comment]
bytes_to_string function works correctly now.  It took a little bit to figure out how to do that the right way.  About the copyright line, I don't know what the etiquette is for that.  Just add my name in under his?

**Comment by Heather** [comment]
> Just add my name in under his?

I think it will be fine enough

**Comment by Heather** [comment]
@forticulous maybe we may need some tests for it either, if this is really useless on windows it should be moved to UNIX_PROGS but I consider corrections may go as additional pull requests

File: src/shred/shred.rs (hunk 2 above)
Comment (doomrobo) [inline src/shred/shred.rs]: This is indeed a race condition, but there is no easy way around it. `fs::rename()` does not return an error when the target name already exists.

**Comment by doomrobo** [comment]
Which section is that?

**Comment by ebfe** [comment]
On Sun, Jan 03, 2016 at 11:20:17AM -0800, Michael Rosenberg wrote:
> Which section is that?

The complete help text.

**Pull request merged**